align/pnr/toplevel.py

```diff
@@ -123,7 +124,69 @@ def route_bottom_up( DB, drcInfo,
                     bounding_box,
                     current_node_ort, idx, lidx, sel,
                     opath, adr_mode, *, PDN_mode, results_name_map, hierarchical_path):
-    raise NotImplementedError( f'route_bottom_up not yet implemented')
+
+    """We are trying to route the lidx-th placement (also the sel-th)
+"""
+
+    assert lidx == sel
+
+    if lidx != 0:
+        raise NotImplementedError( f'route_bottom_up not yet implemented for multiple placements: lidx {lidx} != 0')
+
+    TraverseOrder = DB.TraverseHierTree()
+
+    assert idx == TraverseOrder[-1]
+
+    new_currentnode_idx_d = {}
+
+    for i in TraverseOrder:
+        current_node = DB.CheckoutHierNode(i, sel) # Make a copy
+        DB.hierTree[idx].n_copy += 1
+
+        logger.info( f'Order: {i} {current_node.name}')
+
+        current_node_name = current_node.name
+
+        logger.info( f'Existing parents: {current_node.parent}')
+        current_node.parent = []
+
+        logger.info( f'Existing LL,UR,abs_orient: {current_node.LL.x},{current_node.LL.y} {current_node.UR.x},{current_node.UR.y} {current_node.abs_orient}')
+
+        current_node.LL = bounding_box.LL
+        current_node.UR = bounding_box.UR
+        current_node.abs_orient = current_node_ort
+
+        # doesn't work if I don't do this
+        DB.TransformNode(current_node, current_node.LL, current_node.abs_orient, TransformType.Forward)
+
+        # Remap using new bottom up hNs
+        for bit,blk in enumerate(current_node.Blocks):
+            child_idx = blk.child
+            if child_idx >= 0:
+                # SMB This needs to be here.
+                DB.CheckinChildnodetoBlock(current_node, bit, DB.hierTree[new_currentnode_idx_d[child_idx]])
```

Comment:
@854768750 I probably need to make a copy here (I guess after the first instance), then run `CheckinChildnodetoBlock`

Reply:
No, this seems to be doing more or less the correct thing. CheckinChildnodetoBlock seems to copy some terminals into the parent after translating them, which is what we want.